### PR TITLE
Touchups to the CLI allocation process: support for CSV, bugfixes, etc.

### PR DESCRIPTION
**Pull request by cygnusv**
Command to use the allocation CLI:
```nucypher-deploy allocations --provider <PROVIDER> --poa  --allocation-infile <ALLOCATION_FILE>```

Example of allocation file in CSV:
```
"beneficiary_address","name","amount","duration_seconds"
"0x1715a31490371EBC469D775d6b384C354Be803C7","Aunt Irma",15000000000000000000,259200
```
Note you need those exact column names.

### CODE REVIEW

File: nucypher/blockchain/eth/actors.py [hunk 1]
```diff
@@ -518,11 +533,22 @@ def deploy_beneficiary_contracts(self,
...
+                allocation_data = list()
+                reader = csv.DictReader(allocation_file)
+                for entry in reader:
+                    entry['amount'], entry['duration_seconds'] = int(entry['amount']), int(entry['duration_seconds'])
+                    allocation_data.append(entry)
+            else:
+                data = allocation_file.read()
+                try:
+                    allocation_data = json.loads(data)
+                except JSONDecodeError:
```
Comment (fjarri) [inline nucypher/blockchain/eth/actors.py]: Hm, and ignore any other errors? I understand that this `try` block came from the original version, but perhaps we can just remove it - the CSV part doesn't have anything like that.

File: nucypher/blockchain/eth/actors.py [hunk 2]
```diff
@@ -383,13 +384,20 @@ def deploy_beneficiary_contracts(self,
                                      emitter: StdoutEmitter = None,
                                      ) -> Dict[str, dict]:
         """
-        The allocation file is a JSON file containing a list of allocations. Each allocation has a:
+        The allocation file is a CSV file containing a list of allocations. Each allocation has a:
```
Comment (fjarri) [inline nucypher/blockchain/eth/actors.py]: "A CSV or a JSON file", perhaps? Also, judging by the code, the CSV file is recognized by extension, and anything else is considered JSON - perhaps is worth mentioning too.

File: nucypher/blockchain/eth/actors.py (hunk 1 above)
Comment (cygnusv) [reply nucypher/blockchain/eth/actors.py]: Current catch block doesn't achieve anything, so yes, perhaps it's better to simply remove it.

Comment (cygnusv) [reply nucypher/blockchain/eth/actors.py]: ✅

File: nucypher/blockchain/eth/actors.py (hunk 2 above)
Comment (cygnusv) [reply nucypher/blockchain/eth/actors.py]: ✅

File: nucypher/blockchain/eth/actors.py [hunk 3]
```diff
@@ -518,11 +533,19 @@ def deploy_beneficiary_contracts(self,
     @staticmethod
     def __read_allocation_data(filepath: str) -> list:
         with open(filepath, 'r') as allocation_file:
-            data = allocation_file.read()
-            try:
+            if filepath.endswith(".csv"):
+                reader = csv.DictReader(allocation_file)
+                allocation_data = list(reader)
+            else:  # Assume it's JSON by default
+                data = allocation_file.read()
```
Comment (michwill) [inline nucypher/blockchain/eth/actors.py]: Why not `json.load(allocation_file)`?

Comment (cygnusv) [reply nucypher/blockchain/eth/actors.py]: Why not indeed.

**Comment by cygnusv** [comment]
@michwill I think it definitely falls within the scope of the PR. Let me try to add that too.

**Comment by cygnusv** [comment]
> Not sure if it's within the scope of the PR... But what if we save non-failed allocations after every deployment to some csv file, so that if something terrible happened (CryptoKitty attack, hardware wallet broke down, ...), we can easily continue from the file which contains what's left to deploy?

Oh I just forgot that I already did that! 😅 
It doesn't create a separate file, but instead it reports failed allocations in the summary file (CSV format) which is created at the end of the process. This file cannot be directly used as input to the allocation process (since it also contains successful allocations) but it's a very good start to create a new file by hand. I think the human touch in such moments is important so perhaps it's good to not automate more than necessary.

**Comment by michwill** [comment]
Sure, if you prefer this way.
Also, sometimes geth may report the transaction as "failed" when it succeeds (when there is a chain reorg, for example). It's actually a good idea to check if something is deployed or not already. Not sure if possible though

**Pull request merged**